### ISSUE TITLE
Lint request: filter(is_some).map(unwrap)

### ISSUE THREAD

**Issue by jplatte** [rust-lang/rust-clippy#8920]
### What it does

Lint about usage of `.filter(/* ... */).map(/* ... */)` calls that could be simplified to `.filter_map(/* ... */)`.

### Lint Name

filter_then_map

### Advantage

- Removes an unnecessary `.unwrap()`
  - … or `.expect()`
  - … or `.unwrap_or*` with a never-used default value / expression
- Makes the code easier to read and shorter

### Example

```rust
struct Foo {
    field: Option<String>,
}

fn main() {
    let vec = vec![Foo {
        field: Some("value".to_owned()),
    }];
    let iter = vec
        .iter()
        .filter(|f| f.field.is_some())
        .map(|f| f.field.clone().unwrap());

    for field in iter {
        dbg!(field);
    }
}
```

Could be written as:

```rust
struct Foo {
    field: Option<String>,
}

fn main() {
    let vec = vec![Foo {
        field: Some("value".to_owned()),
    }];
    let iter = vec.iter().filter_map(|f| f.field.clone());

    for field in iter {
        dbg!(field);
    }
}
```

**Comment by mikerite** [comment]
This looks like a [manual_filter_map](https://rust-lang.github.io/rust-clippy/master/index.html#manual_filter_map) false negative to me.
